Source/DafnyCore/Resolver/PreType/PreTypeConstraints.cs

```diff
@@ -179,26 +241,40 @@ public class PreTypeConstraints {
     /// Try to resolve each proxy using its sub-bound constraints (if "fromSubBounds" is "true") or
     /// its super-bound constraints (if "fromSubBounds" is "false"). Add an equality constraint for
     /// any proxy whose head can be determined.
+    ///
+    /// If "ignoreUnknowns" is true, then ignore any constraint where the bound is a proxy or for which the running join/meet computation
+    /// is ill-defined.
+    /// If "ignoreUnknowns" is false, then don't resolve a proxy if it has unknowns.
+    ///
     /// Return "true" if any such equality constraint was added.
     /// </summary>
-    bool TryResolveTypeProxiesUsingKnownBounds(bool fromSubBounds) {
+    bool TryResolveTypeProxiesUsingKnownBounds(bool fromSubBounds, bool ignoreUnknowns) {
       // First, compute the join/meet of the sub/super-bound heads of each proxy
-      Dictionary<PreTypeProxy, TopLevelDecl> candidateHeads = new();
+      Dictionary<PreTypeProxy, TopLevelDecl> candidateHeads = new(); // if !ignoreUnknowns, map to null to indicate the proxy has unknowns
       Dictionary<PreTypeProxy, SubtypeConstraint> constraintOrigins = new();
       foreach (var constraint in unnormalizedSubtypeConstraints) {
         var proxy = (fromSubBounds ? constraint.Super : constraint.Sub).Normalize() as PreTypeProxy;
         var bound = (fromSubBounds ? constraint.Sub : constraint.Super).Normalize() as DPreType;
-        if (proxy != null && bound != null) {
+        if (proxy != null) {
           if (!candidateHeads.TryGetValue(proxy, out var previousBest)) {
-            candidateHeads.Add(proxy, bound.Decl);
-            constraintOrigins.Add(proxy, constraint);
+            // we haven't seen this proxy before
+            if (bound != null) {
+              candidateHeads.Add(proxy, bound.Decl);
+              constraintOrigins.Add(proxy, constraint);
```

Comment:
Factor out that line after conditional for both branches?

Reply:
The `constraintOrigins.Add` happens in 2 of the 3 branches. (Note that this is an `if ... else if`, not an `if ... else`.)